libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/ChannelFactory.java

```diff
@@ -41,22 +42,30 @@ protected ChannelFactory(RawChannelFactory rawChannelFactory) {
         this.rawChannelFactory = rawChannelFactory;
     }
 
-    public Socket openNioChannel(InetSocketAddress remoteAddress, SocketSelector selector) throws IOException {
+    public Socket openNioChannel(InetSocketAddress remoteAddress, Supplier<SocketSelector> supplier) throws IOException {
         SocketChannel rawChannel = rawChannelFactory.openNioChannel(remoteAddress);
+        SocketSelector selector = supplier.get();
         Socket channel = internalCreateChannel(selector, rawChannel);
         scheduleChannel(channel, selector);
         return channel;
     }
 
-    public Socket acceptNioChannel(NioServerSocketChannel serverChannel, SocketSelector selector) throws IOException {
-        SocketChannel rawChannel = rawChannelFactory.acceptNioChannel(serverChannel);
-        Socket channel = internalCreateChannel(selector, rawChannel);
-        scheduleChannel(channel, selector);
-        return channel;
+    public Socket acceptNioChannel(ServerChannelContext serverContext, Supplier<SocketSelector> supplier) throws IOException {
+        SocketChannel rawChannel = rawChannelFactory.acceptNioChannel(serverContext);
+        // Null is returned if there are no pending sockets to accept
+        if (rawChannel == null) {
+            return null;
+        } else {
+            SocketSelector selector = supplier.get();
+            Socket channel = internalCreateChannel(selector, rawChannel);
+            scheduleChannel(channel, selector);
+            return channel;
+        }
     }
 
-    public ServerSocket openNioServerSocketChannel(InetSocketAddress address, AcceptingSelector selector) throws IOException {
+    public ServerSocket openNioServerSocketChannel(InetSocketAddress address, Supplier<AcceptingSelector> supplier) throws IOException {
```

Comment:
why do we need a supplier here can't we create the selector earlier?

Reply:
The reason for the supplier is to avoid skipping selectors (round-robin) when the attempt to open a channel fails.